src/nitsm/pinquerycontexts.py

```diff
@@ -124,3 +133,42 @@ def _publish_bool_2d(self, data, published_data_id):
             return self._tsm_context.Publish_7(self._pins, published_data_id, data)
         else:
             return self._tsm_context.Publish_8(self._pins, published_data_id, data)
+
+    def publish_pattern_results(
+        self, instrument_site_pattern_results: "_PublishPatternArg", published_data_id=""
+    ):
+        """
+        Publishes results from NI-Digital pattern burst to the Semiconductor Multi Test step for all
+        sites in the Semiconductor Module context. Leave the Pin column blank for the test on the
+        Semiconductor Multi Test step when publishing pattern results with this method.
+
+        Args:
+            instrument_site_pattern_results:
+                The pattern result data from multiple pins connected to one or more NI-Digital
+                Pattern instruments. Provide a 1D sequence to publish pattern results from a single
+                NI-Digital Pattern instrument session. Provide a 2D sequence to publish pattern
+                results from multiple NI-Digital Pattern instrument sessions. Each element in the
```

Comment:
This method is intended to publish the results from the [burst_pattern ](https://nimi-python.readthedocs.io/en/master/nidigital/class.html#burst-pattern)method. That method does not return a 1D sequence so to use publish_pattern_results, the customer would have to transform the return values from burst_pattern into 1D sequences. Transforming the data is not trivia because there is no guarantee of the order of the values in the burst_pattern return value. For example, if burst_pattern returns `{1: True, 0: False}` and `tsm_context.pins_to_nidigital_session` returns `site_list: 'site0,site1'` then we need to pass `[False, True]` to publish_pattern_results. 
Here's what the code module would look like:
```
pin_query, sessions, site_lists = tsm_context.pins_to_nidigital_sessions_for_pattern(pins)

combined_pattern_results = []
for session, site_list in zip(sessions, site_lists)
    session_pattern_results = session.sites[site_list].burst_pattern(...)
    combined_pattern_results.append(transform_data_for_publish(site_list, session_pattern_results))

pin_query.publish_pattern_results(combined_pattern_results)
```
Here's what transform_data_for_publish would do:
```
def transform_data_for_publish(site_list, session_pattern_results)
    transformed_data = []
    for site_string in site_list.split(',')
        site_number = int(site_string[4:])
        transformed_data.append(session_pattern_results[site_number])
    return transformed_data
```
It's important that we do not make it easy for the customer to pass incorrect data to publish. Ideally, we would just pass the dictionaries into the publish_pattern_results method, but we would also need to pass in the site_lists to correctly transform the data. Regardless, I think this is my preferred option. The code module would simpler and the data transformation would occur within the publish call itself.

Reply:
Check out the changes in ea7c03a631515edbba23c41690e36e1cf0f7acaa.

I added a **kwargs to the PinQueryContext initializer that will allow us to pass special variables during the pin query.  I pass the site lists to the pin query context so i can properly format the data before publishing it.  now the user can do something like this:

```python
pin_query, sessions, site_lists = tsm_context.pins_to_nidigital_sessions_for_pattern(pins)

pattern_results = [nidigital.sites[site_list].burst_pattern(...) for site_list in site_lists]

pin_query.publish_pattern_results(pattern_results)
```

Reply:
If you approve of this approach, then we still need to update the documentation for the publish method.

Reply:
Very nice solution!

Reply:
I'd like to consider creating a derived class DigitalPatternPinQueryContext that has the site_lists property instead of using the kwargs. The DigitalPatternPinQueryContext would have the publish_pattern_results method. One potential advantage would be error reporting if a customer tries to call publish_pattern_results on a pin query context obtained from anything other than a digital pattern pin query (pins_to_nidigital_sessions_for_pattern). What do you think?

Reply:
I'm good with that, just mad that change.